glasses/nn/att/__init__.py

```diff
@@ -249,6 +250,54 @@ def forward(self, x: Tensor) -> Tensor:
         return x * y
 
 
+class CBAM(nn.Module):
+    def __init__(
+        self, 
+        features: int, 
+        reduction: int = 16,
+        kernel_size: int = 7, 
+        reduced_features: int = None,
+        activation: nn.Module = ReLUInPlace,
+    ):
+        super().__init__()
+        self.reduced_features=(
+            features // reduction if reduced_features is None else reduced_features
+        )
+
+        self.channel_att = nn.Sequential(
+            OrderedDict(
+                {
+                    "fc1": nn.Conv2d(features, self.reduced_features, kernel_size=1, bias=False),
+                    "act1": activation(),
+                    "fc2": nn.Conv2d(self.reduced_features, features, kernel_size=1, bias=False),
+                }
+            )
+        )
+        self.spatial_att = nn.Sequential(
+            OrderedDict(
+                {
+                    "conv": nn.Conv2d(2, 1, kernel_size=kernel_size, padding=kernel_size//2, bias=False),
+                }
+            )
+        )
+        self.gate = nn.Sigmoid()
+    
+    def forward(self, x):
+        # channel attention
+        channel_avg_out = self.channel_att(x.mean((2, 3), keepdim=True))
```

Comment:
can we use `einops` `Reduce` layer instead directly inside self.channel_att. Btw what's the difference between normal `SpatialSE` and this version with conv2d with `kernel_size=1`?

Reply:
Conv2d with `kernel_size=1` equals to a `FC` Layer, these two kind of codes have the same meaning
```
x = torch.randn(1, 512, 16, 16)
fc = nn.Linear(512,  256)
mean_x = x.mean((2,3))
fc(mean_x)
```
```
x = torch.randn(1, 512, 16, 16)
fc = nn.Conv2d(512, 256, kernel_size=1)
mean_x = x.mean((2,3), keepdim=True)
fc(mean_x)
```

Reply:
Yeah! You're wright, it's better to use einops here.